lib/compat/wordpress-6.3/template-parts-preloading.php

```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * Patches resources loaded by the block editor page
+ * to include Template Part posts.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Preloads requests needed for common Template Part posts
+ *
+ * @param array $preload_paths    Preload paths to be filtered.
+ * @param array $context          Context for the preload paths.
+ * @return array    the preload paths.
+ */
+function gutenberg_preload_template_parts( $preload_paths, $context ) {
+
+	// Limit to the Site Editor.
+	if ( ! empty( $context->name ) && 'core/edit-site' !== $context->name ) {
+		return $preload_paths;
+	}
+
+	// Get any template parts defined in theme.json.
+	$theme_json_template_parts = WP_Theme_JSON_Resolver::get_merged_data()->get_template_parts();
```

Comment:
In the plugin, we need to use the classes bundled by Gutenberg: `WP_Theme_JSON_Resolver_Gutenberg` instead of `WP_Theme_JSON_Resolver`.

Reply:
@oandregal On another note it would be nice if there was a stable global function which I could use instead of `::get_merged_data()->get_template_parts()`. Is that a good idea?

Reply:
Absolutely! We should not use the private classes in places other than the high-level functions (`get_global_settings`, etc.). It's tracked as part of https://github.com/WordPress/gutenberg/issues/45171

> Implement wp_get_custom_templates_from_theme: access to customTemplates defined by the theme